### PR TITLE
RUBY-3087 reinstate ability to call insert_many with Enumerable inputs which are not Array instances

### PR DESCRIPTION
**Pull request by midnight-wonderer**
Hello,
As I mentioned on https://github.com/mongodb/mongo-ruby-driver/discussions/2430, we use `insert_many` with an enumerable instead of an array; we want to make sure it will continue to work into the future.

It would be great if you guys could include this test.  
I am happy to adjust the pull request as you see fit.

### CODE REVIEW

**Comment by alexbevi** [comment]
Hi @midnight-wonderer, and thanks for submitting this PR! We will review this and get back to you with any feedback.

**Comment by midnight-wonderer** [comment]
It already breaks D: (test failed)  
Let me see how I can help with the situation; I'll get back to you shortly.

**Comment by midnight-wonderer** [comment]
Updated!  
The breaking change was introduced in commit#5931b9 by @Neilshweky (pinging to request a code review).
The updated commit should still comply with RUBY-2976.  

The user should be responsible for side effects that may occur because of multiple `each` iterations from the driver. It is the power of duck-typing, and also, it is undocumented, after all.

I would like to suggest subclassing `ArgumentError` so people can confidently `rescue` from the error; submitting an empty operation to `bulk_write` has a valid use case, IMO, but probably another time (not in this PR).

**Comment by Neilshweky** [comment]
Hey @midnight-wonderer thanks for the ping! I just added another test, and some docs stating that we now support enumerables in the respective methods. Would you mind taking a look at that?

**Review by Neilshweky — APPROVED** [review]

File: lib/mongo/collection/view/writable.rb [hunk 1]
```diff
@@ -98,8 +98,6 @@ def find_one_and_delete(opts = {})
           end.first&.fetch('value', nil)
         end
 
-        # db['users'].bulk_write([{insert_one: {x: 1}}, {insert_one: {x: 2}}])
```
Comment (Neilshweky) [inline lib/mongo/collection/view/writable.rb:101]: looks like leftover useless code I left in there. Drive-by fix.

**Comment by midnight-wonderer** [comment]
LGTM (Neil's commit)

File: lib/mongo/collection.rb [hunk 2]
```diff
@@ -720,7 +720,7 @@ def insert_many(documents, options = {})
     # @example Execute a bulk write.
     #   collection.bulk_write(operations, options)
     #
-    # @param [ Array<Hash> ] requests The bulk write requests.
+    # @param [ Array<Hash>, Enumerable ] requests The bulk write requests.
```
Comment (p-mongo) [inline lib/mongo/collection.rb:723]:
```suggestion
    # @param [ Array<Hash>, Enumerable<Hash> ] requests The bulk write requests.
```

File: spec/mongo/collection_crud_spec.rb [hunk 3]
```diff
@@ -362,6 +362,42 @@
...
+
+        let(:source_data) do
+          []
+        end
+
+        let(:result) do
+          authorized_collection.insert_many(source_data.lazy)
+        end
+
+        it 'should accepts them without raising an error' do
```
Comment (p-mongo) [inline spec/mongo/collection_crud_spec.rb:393]: The description on this line doesn't match behavior?

File: lib/mongo/bulk_write.rb [hunk 4]
```diff
@@ -359,6 +359,8 @@ def validate_requests!
             end
           end
         end
+      end.tap do
+        raise ArgumentError, "Bulk write requests cannot be empty" if requests_empty
```
[the remark this answers is not in the document]
Comment (midnight-wonderer) [reply lib/mongo/bulk_write.rb:363]: I am pretty sure the error raised here is already an edge case in some of my previous projects, but the error is according to the spec. I just don't know the rationale behind the spec.

File: lib/mongo/collection.rb [hunk 5]
```diff
@@ -692,7 +692,7 @@ def insert_one(document, opts = {})
     # @example Insert documents into the collection.
     #   collection.insert_many([{ name: 'test' }])
     #
-    # @param [ Array<Hash> ] documents The documents to insert.
+    # @param [ Array<Hash>, Enumerable ] documents The documents to insert.
```
[the remark this answers is not in the document]
Comment (midnight-wonderer) [reply lib/mongo/collection.rb:695]: I don't use type in Ruby, so I don't know which one can be generic.
However, if Enumerable can be generic too and `Array` is also an `Enumerable`, we can also just do `Enumerable<Hash>`.  

There are three possibilities forward:
* `Array<Hash>, Enumerable` as Neil suggested  
probably maybe because `Enumerable` is not a generic type? And also not to confuse developers that are new to the concept of Ruby's module.
* `Array<Hash>, Enumerable<Hash>` as you suggested  
because `Enumerable` is actually a generic type; leaving the `Array` type as is for the same reason as above.
* `Enumerable<Hash>`  
if `Enumerable` is a generic type, and we prioritize conciseness.

**Comment by midnight-wonderer** [comment]
I resolved the spec description issue mentioned by p-mongo, hope you don't mind; could you take a look at the other comments? @Neilshweky

File: docs/release-notes.txt [hunk 6]
```diff
@@ -79,6 +79,7 @@ This release includes the following new features:
 - When 0 is given as the max connection pool size, it is now interpreted to
   mean no limit.
 - The default maximum connection pool size has been increased to 20 from 5.
+- Added support for enumerables in bulk write operations.
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply docs/release-notes.txt:82]: Why won't it need a release not even if we're backporting to 2.18?

File: lib/mongo/bulk_write.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongo/bulk_write.rb:363]: Empty enumerators should raise an error to conform with the spec, in my opinion

File: lib/mongo/collection.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongo/collection.rb:695]: I don't have a strong opinion on this, I had just looked for examples elsewhere in the code base and this is what we did. I don't mind switching it...

File: lib/mongo/bulk_write.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongo/bulk_write.rb:363]: OK no prob, the exception can always be rescued by applications.

File: lib/mongo/collection.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongo/collection.rb:695]: I think `Array<Hash> | Enumerable<Hash>` is redundant, my vote is for `Enumerable<Hash>` only.

@midnight-wonderer This notation simply signifies the expected type of values in the container, we haven't magically come up with a way to add generics to Ruby :)

File: docs/release-notes.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (p-mongo) [reply docs/release-notes.txt:82]: Because the break was in 2.18, thus after the fix there wouldn't be any stable branch where the issue exists so it might be confusing to people to try to work out which versions are broken.

But if you wish to have the note, that's fine, I would just phrase it as "Reinstated support for enumerables" or similar since it was already present and unintentionally broken.

[the remark this answers is not in the document]
Comment (Neilshweky) [reply docs/release-notes.txt:82]: oh okay I will remove it

Comment (Neilshweky) [inline docs/release-notes.txt:82]:
```suggestion
```

File: lib/mongo/collection.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongo/collection.rb:695]: That's fine with me

[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongo/collection.rb]: done

**Review by Neilshweky — APPROVED** [review]

**Comment by Neilshweky** [comment]
@p-mongo  this is rfal

**Review by p-mongo — APPROVED** [review]

**Comment by p-mongo** [comment]
@midnight-wonderer Thank you again for the PR and the fix.
